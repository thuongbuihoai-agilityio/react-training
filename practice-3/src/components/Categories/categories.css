.categories {
  margin-top: 50px;
  text-align: center;
}

.categories__title {
  text-transform: uppercase;
  font-weight: bold;
}

.categories__list {
  display: flex;
  justify-content: center;
  margin: 30px 0 40px 0;
  padding: 0;
}

.categories__item {
  list-style: none;
  cursor: pointer;
  margin: 0 30px;
  position: relative;
  color: #696969;
}

.categories__item:hover {
  color: #ebb576;
}

.categories__item::after {
  content: "";
  width: 5px;
  height: 5px;
  border-radius: 50%;
  background-color: #555;
  position: absolute;
  right: -30px;
  top: 45%;
  color: #696969;
}

.categories__item > a {
  color: black;
  text-decoration: none;
}