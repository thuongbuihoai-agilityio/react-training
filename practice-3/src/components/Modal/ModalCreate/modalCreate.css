.form__error {
  font-size: 11px;
  color: var(--bg-delete);
}