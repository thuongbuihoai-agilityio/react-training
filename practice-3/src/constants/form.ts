import { RULES } from "./rules"

const FORM_VALUES = {
  categoryId: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  name: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  price: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  quantity: {
    value: "",
    rules: [RULES.REQUIRED, RULES.NUMBER],
    error: "",
  },
  description: {
    value: "",
    rules: [RULES.REQUIRED],
    error: "",
  },
  images: {
    value: "",
    rules: [],
    error: "",
  }
}

export default FORM_VALUES
