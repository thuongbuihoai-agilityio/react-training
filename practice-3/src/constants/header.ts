const USERS = {
  username: "Hoai Thuong",
};

export { USERS }