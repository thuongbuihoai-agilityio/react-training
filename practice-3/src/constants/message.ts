const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

export { SUCCESS_MSG };
