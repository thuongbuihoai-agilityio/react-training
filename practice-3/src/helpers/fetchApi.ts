import axios from "axios";
import { Product } from "@/types/product";

const get = async (url: string) => {
  try {
    const res = await axios.get(url);
    return res.data;
  } catch(error) {
    console.log(error)
  }
};

const create = async (url: string, newProduct: Product) => {
  try {
    const res = await axios.post(url, newProduct);
    return res;
  }
  catch(error) {
    console.log(error);
  };
}

const update = async (id: string, productEdit: Product) => {
  try {
    const res = await axios.put(id, {...productEdit});
    return res;
  } catch(error) {
      console.log(error);
  };
}

const remove = async (id: string) => {
  try {
    const res = await axios.delete(id);
    return res;
  } catch(error) {
    console.log(error);
  };
}

export {
  get,
  create,
  update,
  remove
}