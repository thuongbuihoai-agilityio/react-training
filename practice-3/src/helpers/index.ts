const setFieldsValue = (fieldsValue, value: string, fieldName: string) => {
  const fieldData = fieldsValue[fieldName];
  return {
    ...fieldsValue,
    [fieldName]: {
      ...fieldData,
      value: value,
      error: ""
    }
  };
}

export { setFieldsValue };
