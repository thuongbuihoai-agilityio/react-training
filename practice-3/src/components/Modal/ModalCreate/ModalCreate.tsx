import React, { useState } from "react";
import { ModalCreateProps } from "@/types/modal";
import { CategoryProps } from "@/types/category";
import getBase64 from "@/helpers/getBase64";
import FORM_VALUES from "@/constants/form";
import { validate } from "@/helpers/validate";
import { setFieldsValue } from "@/helpers/index";
import { Product } from "@/types/product";
import { FormProps } from "@/types/form";
import useSWR, { Key } from "swr";
import { CATEGORIES_URL } from "@/constants/url";
import { get } from "@/helpers/fetchApi";
import "./modalCreate.css";

const ModalCreate: React.FC<ModalCreateProps> = ({
  hideModalCreate,
  createProduct,
}) => {
  const [newProduct, setNewProduct] = useState([]);
  const key: Key = CATEGORIES_URL;
  const fetcher = () => get<Product[]>(CATEGORIES_URL);
  const { data } = useSWR(key, fetcher);
  const [selectedFile, setSelectedFile] = useState([]);
  const [formValues, setFormValues] = useState<FormProps>(FORM_VALUES);

  const handleClearValidate = () => {
    (Object.keys(formValues) as (keyof typeof formValues)[]).map(
      (fieldName) => {
        formValues[fieldName].error = "";
      }
    );
    hideModalCreate();
  };

  const handleCreateProduct = () => {
    const images: string[] = [];
    for (let i = 0; i < selectedFile.length; i++) {
      images.push(selectedFile[i]);
    }
    setFormValues(validate(formValues));
    const temp = (Object.keys(formValues) as (keyof typeof formValues)[]).map(
      (fieldName) => {
        if (formValues[fieldName].error) {
          return false;
        }
      }
    );

    if (!temp.includes(false)) {
      handleClearValidate();
      createProduct({ images, ...newProduct } as unknown as Product);
    }
  };

  const handleChange = (event: { target: { value: string; name: string } }) => {
    const value = event.target.value;
    const fieldName = event.target.name;
    setNewProduct({ ...newProduct, [fieldName]: value });
    setFormValues(setFieldsValue(formValues, value, fieldName));
  };

  const imageChange = async (event: { target: { files: any } }) => {
    const files = event.target.files;
    if (files) {
      for (let i = 0; i < files.length; i++) {
        const imageSrc = await getBase64(files[i]);
        setSelectedFile((selectedFile) => [...selectedFile, imageSrc] as any);
      }
    }
  };

  const handleDeleteImage = (event: { target: Element | any }) => {
    const target = event.target as Element;
    const indexOfArr = selectedFile.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    selectedFile.splice(indexOfArr, 1);
    setSelectedFile([...selectedFile]);
  };

  return (
    <div
      data-testid="modal-create"
      className="modal-create"
      id="bookDeleteModal"
    >
      <div className="modal-dialog-modalUpdate">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-heading" id="productModalLabel">
              Add New Product Information
            </h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name: </label>
              <input
                data-testid="change-value-name"
                className="form__input"
                type="text"
                name="name"
                onChange={handleChange}
              />
              <small className="form__error">
                {formValues.name.error ? formValues.name.error : ""}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Description: </label>
              <textarea
                className="form__text"
                name="description"
                id=""
                cols={30}
                rows={5}
                onChange={handleChange}
              ></textarea>
              <small className="form__error">
                {formValues.description.error
                  ? formValues.description.error
                  : ""}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories: </label>
              <select
                className="form__select"
                name="categoryId"
                onChange={handleChange}
              >
                <option value="">Select category</option>
                {data?.map(({ id, name }: CategoryProps, index: number) => (
                  <option key={index} value={id}>
                    {name}
                  </option>
                ))}
              </select>
              <small className="form__error">
                {formValues.categoryId.error ? formValues.categoryId.error : ""}
              </small>
            </div>
            <div id="form__number" className="form-control">
              <div className="form-control">
                <label htmlFor="">Price: </label>
                <input
                  className="modal__input"
                  type="number"
                  min={0}
                  name="price"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues.price.error ? formValues.price.error : ""}
                </small>
              </div>
              <div className="form-control">
                <label htmlFor="">Quantity: </label>
                <input
                  className="modal__input"
                  type="number"
                  min={0}
                  name="quantity"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues.quantity.error ? formValues.quantity.error : ""}
                </small>
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {selectedFile.length > 0 &&
                  selectedFile.map((src, key: number, index) => (
                    <img
                      data-testid="after-change-file"
                      className="form__img"
                      key={key}
                      data-id={src}
                      src={src}
                      onClick={handleDeleteImage}
                    />
                  ))}
                <input
                  data-testid="change-file"
                  className="form__input--img"
                  type="file"
                  id="file"
                  multiple
                  name="images"
                  onChange={imageChange}
                />
              </div>
            </div>
          </div>
          <div className="modal-footer-modalUpdate">
            <button
              data-testid="hide-modal-btn"
              className="btn btn__no"
              onClick={handleClearValidate}
            >
              Cancel
            </button>
            <button
              data-testid="add-new-product"
              className="btn btn__yes"
              onClick={() => handleCreateProduct()}
            >
              Submit
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ModalCreate;
