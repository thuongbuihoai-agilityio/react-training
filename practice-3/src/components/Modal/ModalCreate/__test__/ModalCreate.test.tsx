import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ModalCreate from "../ModalCreate";
import { useState } from "react";
import mockAxios from "@/__mocks__/axios";
import { CATEGORY_MOCKING_LIST } from "@/constants/categories";
import { CATEGORIES_URL } from "@/constants/url";
import { get } from "@/helpers/fetchApi";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("Modal create component", () => {
  const createProduct = jest.fn();
  const hideModalCreate = jest.fn();
  beforeEach(()=>{
    (useState as jest.Mock).mockImplementation(jest.requireActual("react").useState);
  })

  afterEach(() => {
    mockAxios.reset();
  });

  test("get categories item should call", async () => {
    mockAxios.get.mockResolvedValueOnce({data: CATEGORY_MOCKING_LIST});
    const result = await get(CATEGORIES_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(CATEGORIES_URL);
    expect(result).toEqual(CATEGORY_MOCKING_LIST);
  });

  test("should render modal create component", () => {
    const { getByTestId } = render(<ModalCreate hideModalCreate={() => {}} createProduct={() => {}} />);
    expect(getByTestId("modal-create")).toBeInTheDocument();
  });

  test("should hide modal create when click Cancel", () => {
    const { getByTestId } = render(<ModalCreate hideModalCreate={hideModalCreate} createProduct={() => {}} />);
    const submitBtn = getByTestId("hide-modal-btn");
    fireEvent.click(submitBtn);
    expect(hideModalCreate).toHaveBeenCalled();
  });

  test("should create product when click Submit", () => {
    const { getByTestId } = render(<ModalCreate hideModalCreate={() => {}} createProduct={createProduct} />);
    const submitBtn = getByTestId("submit-btn");
    fireEvent.click(submitBtn);
    expect(hideModalCreate).toHaveBeenCalled();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<ModalCreate hideModalCreate={() => {}} createProduct={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});