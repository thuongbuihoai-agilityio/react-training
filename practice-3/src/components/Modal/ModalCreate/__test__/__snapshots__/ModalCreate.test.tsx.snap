// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal create component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal-create"
    data-testid="modal-create"
    id="bookDeleteModal"
  >
    <div
      class="modal-dialog-modalUpdate"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h4
            class="modal-heading"
            id="productModalLabel"
          >
            Add New Product Information
          </h4>
        </div>
        <div
          class="modal-body"
        >
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Product name: 
            </label>
            <input
              class="form__input"
              data-testid="input"
              name="name"
              type="text"
              value=""
            />
            <small
              class="form__error"
            >
              Field is required
            </small>
          </div>
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Description: 
            </label>
            <textarea
              class="form__text"
              cols="30"
              id=""
              name="description"
              rows="5"
            />
            <small
              class="form__error"
            >
              Field is required
            </small>
          </div>
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Categories: 
            </label>
            <select
              class="form__select"
              name="categoryId"
            >
              <option
                value=""
              >
                Select category
              </option>
            </select>
            <small
              class="form__error"
            >
              Field is required
            </small>
          </div>
          <div
            class="form-control"
            id="form__number"
          >
            <div
              class="form-control"
            >
              <label
                for=""
              >
                Price: 
              </label>
              <input
                class="modal__input"
                data-testid="input"
                min="0"
                name="price"
                type="number"
                value=""
              />
              <small
                class="form__error"
              >
                Field is required
              </small>
            </div>
            <div
              class="form-control"
            >
              <label
                for=""
              >
                Quantity: 
              </label>
              <input
                class="modal__input"
                data-testid="input"
                min="0"
                name="quantity"
                type="number"
                value=""
              />
              <small
                class="form__error"
              >
                Field is required
              </small>
            </div>
          </div>
          <div
            class="form-control"
          >
            <div
              class="form__img--list"
            >
              <input
                class="form__input--img"
                data-testid="input"
                id="file"
                multiple=""
                name="images"
                type="file"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="modal-footer-modalUpdate"
        >
          <button
            class="btn btn__no"
            data-testid="hide-modal-btn"
          >
            Cancel
          </button>
          <button
            class="btn btn__yes"
            data-testid="submit-btn"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
